varcode/variant_collection.py

```diff
@@ -1,122 +1,129 @@
-            self.raw_reference_name, self.records = _load_maf(filename)
+            if reference_path:
+                self.reference_name = infer_reference_name(reference_path)
+            else:
+                raise ValueError(
+                    "Must specify one of reference_path or reference_name")
+
+        if ensembl_release:
+            self.ensembl_release = ensembl_release
         else:
-            raise ValueError("Unrecognized file type: %s" % (filename,))
-
+            self.ensembl_release = ensembl_release_number_for_reference_name(
+                self.reference_name)
 
-        self.filename = filename
-        self.reference_name = infer_reference_name(self.raw_reference_name)
-        self.ensembl_release = ensembl_release_number_for_reference_name(
-            self.reference_name)
         self.annot = VariantAnnotator(ensembl_release=self.ensembl_release)
-
-        if drop_duplicates:
-            filtered_records = []
-            seen = set()
-            for record in self.records:
-                key = record.short_description()
-                if key not in seen:
-                    seen.add(key)
-                    filtered_records.append(record)
-            self.records = filtered_records
+        self.original_filename = original_filename
 
     def __len__(self):
-        return len(self.records)
+        return len(self.variants)
 
     def __iter__(self):
-        return iter(self.records)
+        return iter(self.variants)
 
     def __str__(self):
-        s = "VariantCollection(filename=%s, reference=%s)" % (
-            self.filename, self.reference_name)
-        for record in self.records:
-            s += "\n\t%s" % record
+        fields = [
+            ("n_variants", len(self.variants)),
+            ("reference", self.reference_name)
+        ]
+
+        if self.original_filename:
+            fields.append(("filename", self.original_filename))
+
+        s = "VariantCollection(%s)" % (
+            ", ".join(
+                "%s=%s" % (k,v) for (k,v) in fields))
+        for variant in self.variants:
+            s += "\n\t%s" % variant
         return s
 
     def __repr__(self):
         return str(self)
 
+    def clone(self, new_variants=None):
+        """
+        Create copy of VariantCollection with same metadata but possibly
+        different Variant entries. If no variants provided, then just make a
+        copy of self.variants.
+        """
+        if new_variants is None:
+            new_variants = self.variants
+        return VariantCollection(
+            variants=list(new_variants),
+            original_filename=self.original_filename,
+            reference_path=self.reference_path,
+            reference_name=self.reference_name,
+            ensembl_release=self.ensembl_release)
+
+    def drop_duplicates(self):
+        """
+        Create a new VariantCollection without any duplicate variants.
+        """
+        return self.clone(set(self.variants))
```

Comment:
Oh, I thought there already was one! I guess I forgot to add it when I adding `__hash__` and `__eq__` to all the PyEnsembl objects.